### PR TITLE
Improve code quality of ExternalTaskSensor

### PR DESCRIPTION
**Pull request by turbaszek**
This PR adds type hints and refactors slightly the code of ExternalTaskSensor and ExternalTaskMarker

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/sensors/external_task_sensor.py [hunk 1]
```diff
@@ -203,35 +204,21 @@ def get_count(self, dttm_filter, session, states):
...
-                    DR.dag_id == self.external_dag_id,
-                    DR.state.in_(states),  # pylint: disable=no-member
-                    DR.execution_date.in_(dttm_filter),
-                )
-                .scalar()
-            )
-        return count
 
-    def _handle_execution_date_fn(self, context):
+        # .count() is inefficient
```
Comment (kaxil) [inline airflow/sensors/external_task_sensor.py]: Misplaced comment, I think this needs to be above L207

**Review by kaxil — APPROVED** [review]
1 minor comment, else LGTM

File: airflow/sensors/external_task_sensor.py [hunk 2]
```diff
@@ -203,35 +204,21 @@ def get_count(self, dttm_filter, session, states):
...
-            count = (
-                session.query(func.count())
-                .filter(
-                    TI.dag_id == self.external_dag_id,
-                    TI.task_id == self.external_task_id,
-                    TI.state.in_(states),  # pylint: disable=no-member
-                    TI.execution_date.in_(dttm_filter),
-                )
-                .scalar()
+            qry = qry.filter(
```
Comment (XD-DENG) [inline airflow/sensors/external_task_sensor.py]: I may have missed or mis-understood something: the original code is filtering against `TaskInstance` and `DagRun` separately in different conditions. But this new version is filtering against `DagRun` or `TaskInstance` PLUS `DagRun`.

But I'm simply comparing the lines, so as I said above, possibly I misunderstood something.

Comment (turbaszek) [reply airflow/sensors/external_task_sensor.py]: Yes, my bad - I should read the code more thoughtfully! 🤯

Comment (XD-DENG) [reply airflow/sensors/external_task_sensor.py]: No worries at all ;-)

File: airflow/sensors/external_task_sensor.py [hunk 3]
```diff
@@ -149,38 +148,22 @@ def poke(self, context, session=None):
             dttm = context['execution_date']
 
         dttm_filter = dttm if isinstance(dttm, list) else [dttm]
```
Comment (turbaszek) [inline airflow/sensors/external_task_sensor.py:150]: Btw. this is something that bothers me. We "believe" in operator atomicity but this functionality allows users to provide a function that returns a list of dates. If list is provided then we query not for single DAG/task but for multiple ones.

But users will get 1/0 response - task/dag failed. there will be not information how many instances failed or for which dates. I think that the proper Airflow approach would be to have multiple sensors each checking single exec date.

@kaxil @XD-DENG WDYT?

File: airflow/sensors/external_task_sensor.py (hunk 2 above)
Comment (kaxil) [reply airflow/sensors/external_task_sensor.py]: eeks -- Good catch @XD-DENG .

File: airflow/sensors/external_task_sensor.py (hunk 3 above)
Comment (XD-DENG) [reply airflow/sensors/external_task_sensor.py:150]: The use case I can think of to justify this setting-up/design is that a user is checking & depending a series of external DAG/Task.  For example, I want to depend on DAG A, and what I want to ensure is DAG A has succeeded in the past a few consecutive days (or maybe more complex case, not consecutive). Using Multiple Sensors may multiply the resource usage and may make the DAG authoring more complex in this scenario?

From another perspective, if someone prefers to have multiple sensors each checking single exec date, the current design can meet his/her requirement as well.

So personally this doesn't bother me (but as you mentioned, not giving information about how many TI failed or failed for which dates is not nice enough)

Comment (kaxil) [reply airflow/sensors/external_task_sensor.py:150]: Create a back-log ticket maybe to improve the design further or a flag to have one or the other option

Comment (turbaszek) [reply airflow/sensors/external_task_sensor.py:150]: Created https://github.com/apache/airflow/issues/12577

**Review by XD-DENG — APPROVED** [review]

**Pull request merged**